File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {

    public void dfs(int node,boolean[] vis,List<List<Integer>> adj)
    {
        vis[node] = true;

        for(Integer val : adj.get(node))
        {
            if(!vis[val])
            {
                dfs(val,vis,adj);
            }
        }
    }


    public int findCircleNum(int[][] isConnected) {
        int vertices = isConnected.length;

        List<List<Integer>> adj = new ArrayList<>();

        for(int i = 0; i <= vertices; i++)
        {
            adj.add(new ArrayList<Integer>());
        }

        int n = isConnected.length;
        int m = isConnected[0].length;

        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                // if(i != j)
                // {
                    if(isConnected[i][j] == 1)
                    {
                        int u = i+1;
                        int v = j+1;
                        adj.get(u).add(v);
                        adj.get(v).add(u);
                    }
                // }
            }
        }

        boolean[] vis = new boolean[vertices+1];
        int cnt = 0;

        for(int i = 1; i <= vertices; i++)
        {
            if(!vis[i])
            {
                cnt++;
                dfs(i,vis,adj);
            }
        }

        return cnt;
        
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:

    void dfsFunc(int node, int vis[], vector<int> &dfs, vector<int> adj[])
    {
        vis[node] = 1;
        for(auto it : adj[node])
        {
            if(vis[it] != 1)
            {
                dfsFunc(it,vis,dfs,adj);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int v = isConnected.size();
        int vis[v+1];
        for(int i = 0; i < (v+1); i++)
        {
            vis[i] = 0;
        }

        vector<int> adj[v+1];

        for(int i = 0; i < isConnected.size(); i++)
        {
            for(int j = 0; j < isConnected.size(); j++)
            {
                if(i != j && isConnected[i][j] == 1)
                {
                    int u = i + 1;
                    int v = j + 1;
                    adj[u].push_back(v);
                    adj[v].push_back(u);
                }
            }
        }

        int cnt = 0;
        vector<int> dfs;
        
        for(int node = 1; node <= v; node++)
        {
            if(vis[node] != 1)
            {
                dfsFunc(node,vis,dfs,adj);
                cnt++;
            }
        }

        return cnt;
    }
};